RecoMuon/MeasurementDet/src/MuonDetLayerMeasurements.cc

```diff
@@ -150,34 +150,51 @@ MuonRecHitContainer MuonDetLayerMeasurements::recHits(const GeomDet* geomDet,
-
-	if (chamberId.layer() == 0){
-	  // Get the GEMSegment which relies on this chamber
-	  GEMSegmentCollection::range range = theGEMSegments->get(chamberId);
-	  LogDebug("Muon|RecoMuon|MuonDetLayerMeasurements") << "Number of GEM rechits available =  " << theGEMSegments->size()
-							     <<", from chamber: "<< chamberId<<std::endl;
-	  // Create the MuonTransientTrackingRecHit
-	  for (GEMSegmentCollection::const_iterator rechit = range.first;
-	       rechit!=range.second; ++rechit){
-	    LogDebug("Muon|RecoMuon|MuonDetLayerMeasurements") << "On GEM iteration " <<std::endl;
-	    result.push_back(MuonTransientTrackingRecHit::specificBuild(geomDet,&*rechit));
-	  }
+	if (chamberId.layer() != 0)
+	  return result;
+	// return all segments and rechits not used in segments
+	// Get the GEMSegment which relies on this chamber
+	GEMSegmentCollection::range seg_range = theGEMSegments->get(chamberId.superChamberId());
+	LogDebug("Muon|RecoMuon|MuonDetLayerMeasurements") << "Number of GEM rechits available =  " << theGEMSegments->size()
+							   <<", from chamber: "<< chamberId<<std::endl;
+	// Create the MuonTransientTrackingRecHit
+	for (GEMSegmentCollection::const_iterator segment = seg_range.first;
+	     segment!=seg_range.second; ++segment){
+	  LogDebug("Muon|RecoMuon|MuonDetLayerMeasurements") << "On GEM iteration " <<std::endl;
+	  result.push_back(MuonTransientTrackingRecHit::specificBuild(geomDet,&*segment));
 	}
-	else {
-	  // Get the GEMRecHit which relies on this chamber
-	  GEMRecHitCollection::range range = theGEMRecHits->get(chamberId);
-	  LogDebug("Muon|RecoMuon|MuonDetLayerMeasurements") << "Number of GEM rechits available =  " << theGEMRecHits->size()
+	
+	// Get the GEMRecHits not used in GEMSegment
+	for (int layer = 1; layer < 3; ++layer){
+	  for(int roll = 1; roll <= GEMDetId::maxRollId; ++roll){
+	    GEMRecHitCollection::range range = theGEMRecHits->get( GEMDetId(chamberId.region(),1,chamberId.station(),layer,chamberId.chamber(),roll) );
+	    LogDebug("Muon|RecoMuon|MuonDetLayerMeasurements") << "Number of GEM rechits available =  " << theGEMRecHits->size()
 							     <<", from chamber: "<< chamberId<<std::endl;
-	  // Create the MuonTransientTrackingRecHit
-	  for (GEMRecHitCollection::const_iterator rechit = range.first;
-	       rechit!=range.second; ++rechit){
-	    LogDebug("Muon|RecoMuon|MuonDetLayerMeasurements") << "On GEM iteration " <<std::endl;
-	    result.push_back(MuonTransientTrackingRecHit::specificBuild(geomDet,&*rechit));
+	    for (GEMRecHitCollection::const_iterator rechit = range.first;
+		 rechit!=range.second; ++rechit){
+
+	      const GEMRecHit *gemRecHit = (&*rechit);
+	      bool isUsed = false;
+	      for (GEMSegmentCollection::const_iterator segment = seg_range.first;
+		   segment!=seg_range.second; ++segment){
+		for (auto seg : segment->recHits()){
+		  const GEMRecHit * segrechit = dynamic_cast<const GEMRecHit *>(seg);
+		  if (segrechit == gemRecHit){
```

Comment:
this comparison of pointers will not work because GEMSegment keeps a copy of the rechit.
So, it looks like this method ``` MuonDetLayerMeasurements::recHits ``` still returns both segments and hits.
this should be ```(* segrechit == *gemRecHit) ```

[I checked by adding an assert: this condition is always false now and none of the hits are recognized as used.]